### PR TITLE
Add experimental riscv64 builds

### PR DESCRIPTION
**Pull request by kxxt**
With a performance issue(https://github.com/riscv-forks/electron/issues/1) fixed, I think the riscv64 port of electron is mostly ready for VSCodium. There are still some corner cases that lead to segfault but it's already suitable for usage on real riscv64 hardware.

This PR add riscv64 builds to vscodium.

Official electron doesn't support riscv64 so I used a fork maintained by myself.

BTW I noticed some of the patches modifies the compiled javascript files. I think it's cleaner to directly modify the typescript source file and compile ts files to javascrpt.

I haven't implemented riscv64 support in sysroot scripts so this PR only covers a basic riscv64 build, without DEB and RPM targets.

Here's a screenshot:

VSCodium 1.90 running on 64 core SG2042.

The window on the right shows using open-remote-ssh plugin to connect to x86_64 host from riscv64.

Screenshot_20_10_1

### CODE REVIEW

**Comment by lex-ibm** [comment]
Ups... I think we were working on more or less the same...

**Comment by lex-ibm** [comment]
Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?

**Comment by daiyam** [comment]
> Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?

We should split the `ppc64le-and-riscv64-support.patch`.

File: build.sh [hunk 1]
```diff
@@ -14,6 +14,11 @@ if [[ "${SHOULD_BUILD}" == "yes" ]]; then
 
   export NODE_OPTIONS="--max-old-space-size=8192"
 
+  # Some patches(e.g. remove-mangle.patch) causes unused variables and parameters
+  sed -ri '/noUnused(Local|Parameter)s/d' build/tsconfig.json
+  # Recompile ts files in ./build since the patches modifies them
+  yarn --cwd build compile
```
Comment (daiyam) [inline build.sh:20]: No needed since the patches also modify the `.js`

File: patches/remove-mangle.patch [hunk 2]
```diff
@@ -1,37 +1,10 @@
-diff --git a/build/lib/compilation.js b/build/lib/compilation.js
```
Comment (daiyam) [inline patches/remove-mangle.patch:1]: Do not modify that patch.

File: patches/ppc64le-and-riscv64-support.patch [hunk 3]
```diff
@@ -372,10 +393,10 @@ index 4bec13d6e86..3debda1224a 100644
  			Platform::DarwinARM64 => "DarwinARM64",
  			Platform::WindowsX64 => "WindowsX64",
 diff --git a/cli/src/util/prereqs.rs b/cli/src/util/prereqs.rs
-index 20a5bc94b37..50678354245 100644
+index 5d4953acbb2..f22ff60e391 100644
```
Comment (daiyam) [inline patches/ppc64le-and-riscv64-support.patch:396]: Not needed

File: patches/ppc64le-and-riscv64-support.patch [hunk 4]
```diff
@@ -409,10 +430,10 @@ index 079557869e3..c057746b870 100644
  
  if [ "$OS_ID" != "alpine" ]; then
 diff --git a/src/vs/platform/extensionManagement/common/extensionManagement.ts b/src/vs/platform/extensionManagement/common/extensionManagement.ts
-index 9dae82eba07..94246285b26 100644
+index 183f2582871..2e6614cad59 100644
```
Comment (daiyam) [inline patches/ppc64le-and-riscv64-support.patch:433]: Not needed

**Comment by lex-ibm** [comment]
> > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> 
> We should split the `ppc64le-and-riscv64-support.patch`.

Roger that!

File: patches/linux/use-node16.patch [hunk 5]
```diff
@@ -1,19 +1,27 @@
...
- // Based on https://source.chromium.org/chromium/chromium/src/+/main:build/linux/sysroot_scripts/install-sysroot.py.
+diff --git a/build/lib/fetch.ts b/build/lib/fetch.ts
+index dc1de777e04..e62db40bd75 100644
+--- a/build/lib/fetch.ts
++++ b/build/lib/fetch.ts
+@@ -9,6 +9,7 @@ import * as log from 'fancy-log';
+ import * as ansiColors from 'ansi-colors';
+ import * as crypto from 'crypto';
+ import * as through2 from 'through2';
++const fetch: any = require('node-fetch');
```
Comment (daiyam) [inline patches/linux/use-node16.patch:9]: Please revert the changes on the file

File: patches/fix-build-linux.patch [hunk 6]
```diff
@@ -18,20 +18,6 @@ index 8c2b62f..843e8d4 100644
 -	const buildRpmTask = task.define(`vscode-linux-${arch}-build-rpm`, buildRpmPackage(arch));
 +	const buildRpmTask = task.define(`vscode-linux-${arch}-build-rpm`, task.series(prepareRpmTask, buildRpmPackage(arch)));
  	gulp.task(buildRpmTask);
-diff --git a/build/linux/dependencies-generator.js b/build/linux/dependencies-generator.js
```
Comment (daiyam) [inline patches/fix-build-linux.patch:21]: Please revert the changes on the file

File: patches/ext-from-gh.patch [hunk 7]
```diff
@@ -1,22 +1,15 @@
-diff --git a/build/lib/builtInExtensions.js b/build/lib/builtInExtensions.js
-index 1b0adc4..b595123 100644
---- a/build/lib/builtInExtensions.js
-+++ b/build/lib/builtInExtensions.js
```
Comment (daiyam) [inline patches/ext-from-gh.patch:4]: Please revert the changes on the file

**Comment by kxxt** [comment]
> > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> 
> We should split the `ppc64le-and-riscv64-support.patch`.

I think it might be cleaner to keep one patch. If we split the patch, the ppc64 and riscv64 patches both modifies the same line in a file, which will lead to a conflict. WDYT?

**Comment by kxxt** [comment]
> Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?

Does the PPC64LE has an electron fork that publishes electron binaries? If so I could add it in this PR?

File: build.sh (hunk 1 above)
Comment (kxxt) [reply build.sh]: Done

**Comment by lex-ibm** [comment]
> > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> 
> Does the PPC64LE has an electron fork that publishes electron binaries? If so I could add it in this PR?

We don't publish the binaries, but we use the `ELECTRON_MIRROR` variable to set the location for it. The PR that I opened only makes the changes needed to enable the build to happen. Is up to the user to specify where that electron binary comes from.

**Comment by kxxt** [comment]
> > > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> > 
> > 
> > Does the PPC64LE has an electron fork that publishes electron binaries? If so I could add it in this PR?
> 
> We don't publish the binaries, but we use the `ELECTRON_MIRROR` variable to set the location for it. The PR that I opened only makes the changes needed to enable the build to happen. Is up to the user to specify where that electron binary comes from.

Oh sorry, I didn't see your PR. I thought you haven't opened it. About the patches, do you think it's more convenient to split riscv64 and ppc64 or keep one patch?  I think splitting the patch doesn't actually solve the conflict but makes it more complex because we need to stack a patch on top of another.

**Comment by lex-ibm** [comment]
> I think it might be cleaner to keep one patch. If we split the patch, the ppc64 and riscv64 patches both modify the same line in a file, which will lead to a conflict. WDYT?

I kinda agree, I tried splitting the patch and the only way I got it to work was if I based one on top of the other. Also, in the other PR there are some changes to how to get the vscode-sysroot that make it so that we don't need to check for a specific arch.

While developing it I was trying to make it so that porting to RISCV64 was more or less easy, but never tried to build it.

**Comment by lex-ibm** [comment]
>  think splitting the patch doesn't actually solve the conflict but makes it more complex because we need to stack a patch on top of another.

Agree, took me a couple of hours just to split the changes and always diffing everything to make sure I didn't mess anything up. I kinda feel like this could be a single PR or at least we should both coordinate on how to get them both in.

I do recommend you take the patch to `install-sysroot.js`, that makes it easier if the tag changes or we want to use another repo.

**Comment by kxxt** [comment]
> I do recommend you take the patch to `install-sysroot.js`, that makes it easier if the tag changes or we want to use another repo.

I agree with you. There's no need to maintain two methods of using a custom electron repo.

**Comment by kxxt** [comment]
Another way is to decommonize the intermediate vscode artifact, that is, building an intermediate vscode artifact for each architecture. This way we can apply different patches to different architectures. But I don't think it worth the effort for just two architectures that need patching. This method might make more sense if we have patches for four or more architectures that need patching.

**Comment by lex-ibm** [comment]
> This method might make more sense if we have patches for four or more architectures that need patching.

As long as no one asks me to build for `s390x` I'm fine with that.

Let's get started then! I'm in `GMT-5`, so this is probably the best time to contact me.

I see that you replace the `electron.txt` checksums. That would collide with my needs, since that checksums file wouldn't have the sha256 for our binary.

File: package_linux_bin.sh [hunk 8]
```diff
@@ -16,6 +16,21 @@ cd vscode || { echo "'vscode' dir not found"; exit 1; }
...
+  export ELECTRON_SKIP_BINARY_DOWNLOAD=1
+  export PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1
+  # Official electron doesn't support riscv64, adding the checksums of forked electron here
+  # Electron riscv fork might do multiple builds for a same version,
+  # the tag is in the format of v<version>.riscv<build_number>.
+  # In order to be compatible with upstream release scheme, riscv fork also re-upload the latest
+  # RISC-V release to the upstream release tag(v<version>). 
+  # This might be problematic if there is an old electron zip cached locally.
+  rm -f ~/.cache/electron/*/electron-*-linux-riscv64.zip
+  ELECTRON_VERSION="v$(yarn config get target)"
```
Comment (daiyam) [inline package_linux_bin.sh:29]: It would be better to have specific version number and its checksum.
This way, we know for sure which electron version has been used for a specific VSCodium version.
It allows us to be able to reproduce that version if needed.

File: patches/ppc64le-and-riscv64-support.patch [hunk 9]
```diff
@@ -97,10 +97,29 @@ index 6f8144b0954..8a8195e9b66 100644
...
 +++ b/build/gulpfile.vscode.js
-@@ -429,6 +429,8 @@ const BUILD_TARGETS = [
+@@ -339,11 +339,17 @@ function packageTask(platform, arch, sourceFolderName, destinationFolderName, op
+ 			all = es.merge(all, shortcut);
+ 		}
+ 
++		const electronOverride = {};
++		if (platform === 'linux' && arch === 'riscv64') {
++			// electron doesn't support riscv64, so a fork is used here.
++			electronOverride.repo = 'riscv-forks/electron-riscv-releases';
```
Comment (lex-ibm) [inline patches/ppc64le-and-riscv64-support.patch:110]: This can be achieved with `ELECTRON_MIRROR`, can it not?

Comment (kxxt) [reply patches/ppc64le-and-riscv64-support.patch:110]: Yes.

File: package_linux_bin.sh (hunk 8 above)
Comment (kxxt) [reply package_linux_bin.sh:29]: > It would be better to have specific version number and its checksum. This way, we know for sure which electron version has been used for a specific VSCodium version. It allows us to be able to reproduce that version if needed.

Yes, I agree with you. I could use a specific version like v29.4.2.riscv2 but it would require me to make more PRs in the future to maintain the riscv64 revision so I didn't take this approach on the first try.

**Comment by kxxt** [comment]
> > This method might make more sense if we have patches for four or more architectures that need patching.
> 
> As long as no one asks me to build for `s390x` I'm fine with that.
> 
> Let's get started then! I'm in `GMT-5`, so this is probably the best time to contact me.
> 
> I see that you replace the `electron.txt` checksums. That would collide with my needs, since that checksums file wouldn't have the sha256 for our binary.

It wouldn't collide with your needs. The riscv64 and ppc64le pipelines are ran separately so replacing it in riscv64 shouldn't affect it's usage in ppc64le.

File: patches/ppc64le-and-riscv64-support.patch (hunk 9 above)
Comment (daiyam) [reply patches/ppc64le-and-riscv64-support.patch:110]: Not sure. `ELECTRON_MIRROR` seems to be a `electron-builder` variable.
VSCode seems to use `electronRepository` in the `product.json`

https://github.com/microsoft/vscode/blob/0fd9a006f7abcb48ef1e5a9f9ceeca33e1d1daaa/build/lib/electron.ts#L201

Comment (kxxt) [reply patches/ppc64le-and-riscv64-support.patch:110]: > This can be achieved with `ELECTRON_MIRROR`, can it not?

@lex-ibm I didn't see `ELECTRON_MIRROR` anywhere in your PR? Where did you implemented it in your PR?

Comment (kxxt) [reply patches/ppc64le-and-riscv64-support.patch:110]: > This can be achieved with `ELECTRON_MIRROR`, can it not?

@lex-ibm I didn't see `ELECTRON_MIRROR` anywhere in your PR? Where did you implemented it in your PR?

Comment (kxxt) [reply patches/ppc64le-and-riscv64-support.patch:110]: Oh the network glitch caused me to send the comment twice. Sorry.

**Comment by daiyam** [comment]
> It wouldn't collide with your needs. The riscv64 and ppc64le pipelines are ran separately so replacing it in riscv64 shouldn't affect it's usage in ppc64le.

It will be better if you could just patch the `electron.txt`

**Comment by daiyam** [comment]
> > It wouldn't collide with your needs. The riscv64 and ppc64le pipelines are ran separately so replacing it in riscv64 shouldn't affect it's usage in ppc64le.
> 
> Is it ok with you if the binaries for the electron binary and it's sha256 change without VSCodium knowing? I kinda feel like that might be a bit risky.

I agree

File: patches/ppc64le-and-riscv64-support.patch (hunk 9 above)
Comment (kxxt) [reply patches/ppc64le-and-riscv64-support.patch]: Resolved by using a `VSCODE_ELECTRON_REPO` environment variable for the electron repo.

**Comment by daiyam** [comment]
> I actually don't replace the checksums but append to it. @daiyam Does it look good to you? I consider it as a method of "patch".

No, I prefer to have version and checksum in clear text in the repo. There are several benefit to do this:
- every body can look at the code and know which version has been used
- reproducible binaries
- the changes of version are tracked

We do the same for VSCode ([stable.json](https://github.com/VSCodium/vscodium/blob/master/stable.json))

**Comment by lex-ibm** [comment]
I don't know if it helps, but I do remember seeing an environment variable to change the electron version, I just don't remember where. I'll look for it.
With that I think you could use `ELECTRON_MIRROR` and whatever the other variable is and all that would be needed is the patch for `electron.txt`.

**Comment by kxxt** [comment]
> > I actually don't replace the checksums but append to it. @daiyam Does it look good to you? I consider it as a method of "patch".
> 
> No, I prefer to have version and checksum in clear text in the repo. There are several benefit to do this:
> 
>     * every body can look at the code and know which version has been used
> 
>     * reproducible binaries
> 
>     * the changes of version are tracked
> 
> 
> We do the same for VSCode ([stable.json](https://github.com/VSCodium/vscodium/blob/master/stable.json))

Thanks for clarifying. I have squashed the commits and rebased my PR. Will mark ready for review once the CI passes here: https://github.com/kxxt/vscodium/actions/runs/9468203671

**Comment by lex-ibm** [comment]
[Found it!](https://github.com/electron/get/blob/main/src/artifact-utils.ts) Can even specify a custom file name! Does that help?

**Comment by daiyam** [comment]
> @daiyam I write the clear text electron checksum in `package_linux_bin.sh`. Do you think I need to create a patch file for it or it's fine to maintain the checksum in `package_linux_bin.sh`?

You don't need a patch

**Comment by kxxt** [comment]
> > @daiyam I write the clear text electron checksum in `package_linux_bin.sh`. Do you think I need to create a patch file for it or it's fine to maintain the checksum in `package_linux_bin.sh`?
> 
> You don't need a patch

Then I think the PR is ready for review now. The CI is green here(for riscv64 and ppc64le): https://github.com/kxxt/vscodium/actions/runs/9468203671/job/26084811222

**Comment by daiyam** [comment]
> [Found it!](https://github.com/electron/get/blob/main/src/artifact-utils.ts) Can even specify a custom file name! Does that help?

@lex-ibm VSCode seems to use the package [@vscode/gulp-electron](https://www.npmjs.com/package/@vscode/gulp-electron) to download electron (https://github.com/microsoft/vscode/blob/5adc4dcb025aac567c55100c4f2777614644de89/build/lib/electron.ts#L208)

File: package_linux_bin.sh [hunk 10]
```diff
@@ -22,6 +22,17 @@ if [[ "${VSCODE_ARCH}" == "ppc64le" ]]; then
   export VSCODE_SYSROOT_PREFIX='-glibc-2.28'
 fi
 
+if [[ "${VSCODE_ARCH}" == "riscv64" ]]; then
+  export VSCODE_ELECTRON_REPO='riscv-forks/electron-riscv-releases'
+  export ELECTRON_SKIP_BINARY_DOWNLOAD=1
+  export PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1
+  ELECTRON_VERSION="v$(yarn config get target)"
```
Comment (daiyam) [inline package_linux_bin.sh:29]: `ELECTRON_VERSION="v29.4.0"` should be enough

Comment (kxxt) [reply package_linux_bin.sh]: Fixed

File: patches/ppc64le-and-riscv64-support.patch [hunk 11]
```diff
@@ -97,10 +97,33 @@ index 6f8144b0954..8a8195e9b66 100644
...
 +++ b/build/gulpfile.vscode.js
-@@ -429,6 +429,8 @@ const BUILD_TARGETS = [
+@@ -339,11 +339,21 @@ function packageTask(platform, arch, sourceFolderName, destinationFolderName, op
+ 			all = es.merge(all, shortcut);
+ 		}
+ 
++		const electronOverride = {};
++		if (process.env.VSCODE_ELECTRON_REPO) {
++			// electron doesn't support riscv64 and ppc64le, override the repo with `VSCODE_ELECTRON_REPO`.
++			electronOverride.repo = process.env.VSCODE_ELECTRON_REPO;
```
Comment (daiyam) [inline patches/ppc64le-and-riscv64-support.patch:110]: Why not directly modifying the `config` variable?

**Comment by daiyam** [comment]
LGTM

@kxxt Is it ok for you?

**Comment by daiyam** [comment]
@lex-ibm Do you have any feedback?

**Comment by kxxt** [comment]
> LGTM
> 
> @kxxt Is it ok for you?

Yes

**Comment by lex-ibm** [comment]
LGTM! Awesome job!

**Comment by daiyam** [comment]
@kxxt Thanks!

**Comment by kxxt** [comment]
Thanks both of you for helping me to get this PR merged!

About automatically updating the riscv64 electron version and checksum. I think we can implement it in a script and run it in the *-spearhead workflows.
